### ISSUE TITLE
IPA on baremetal server is unable to reach ironic's endpoint - ENETUNREACH

### ISSUE THREAD

**Issue by agewagra** [metal3-io/baremetal-operator#1664]
**Steps to reproduce the bug:**
follow the deployment steps for ironic and baremetal operator. Use the metal3 ironic-image for deploying ironic.  https://book.metal3.io/quick-start.
After deployment, create a baremetalhost for a baremetal server using the yaml file.

**Environment:**
Deployed on an openshift cluster v4.13 - dual stack environment.

**What did you expect to happen:**
The IPA on RAMdisk of baremetal server should reach ironic's endpoint in prebooting of the inspection phase and BMH status should be available.

**Anything else you would like to add:**
We are using a dual stack environment on openshift to make deployments. The baremetal server that we want to provision needs the virtual boot image from ironic to be hosted on ipv6 address. So we gave an ipv6 address for ironic_external_ip. Rest all are ipv4 addresses.

- Baremetal Operator version: latest
- Environment (metal3-dev-env or other): Openshift kubernetes cluster - 4.13

**Bug:**
IPA is trying to perform lookup for node on ironic during inspection. But it is unable to reach the end point. 
<img width="1424" alt="Screenshot 2024-04-11 at 2 51 44 PM" src="https://github.com/metal3-io/baremetal-operator/assets/74589082/32a1c7d5-60c7-4a05-8efc-2d62245465f1">

**Comment by matthewei** [comment]
You should set ironic-endpoint "https://ip:port"

**Comment by Shreshta24** [comment]
Hi,
this is the ironic endpoint that we are using: http://10.59.52.241:6385, which I think is the same as what you are saying.
Correct me if I am wrong

**Comment by agewagra** [comment]
Hi, thanks for the response.
Yeah we are using Openshift as kuberenetes. It is a requirement and yes we are using the latest versions od BMO, Ironic and IPA. 

In the iso image on ramdisk, we confirmed that network data is added in openstack/latest/network_data.json file. Our environment does not support DHCP. We used preprovisioningNetworkData in BMH to add network data. This is supposed to work in DHCP less environment according to https://github.com/metal3-io/metal3-docs/blob/main/design/baremetal-operator/image-builder-integration.md. But the IPA is still waiting for DHCP to assign IP, despite giving the static network data. And since there is no ip that is being assigned to the BareMetalServer, it is unable to reach back ironic's endpoint.

Is there any way to disable waiting for DHCP from IPA?

**Comment by dtantsur** [comment]
Is it a duplicate of https://github.com/metal3-io/metal3-docs/issues/416 then? Could you close one of them?

**Comment by Rozzii** [comment]
@agewagra Have you built your own custon IPA or are you using the default from upstream?
The default upstream IPA doesn't contain the tooling required for network configuration so you would need most likely "Glean" or Cloud-init. 
/triage needs-information

**Comment by agewagra** [comment]
Yeah we were using default, but realised we should build the custom IPA.
Can we build a custom IPA with RHCOS as base image?

**Comment by Rozzii** [comment]
> Yeah we were using default, but realised we should build the custom IPA. Can we build a custom IPA with RHCOS as base image?

Technically by creating custom "elements" for the IPA builder or for OpenStack disk image builder you can build IPA out of any base image. IPA builder is based on Openstack Disk Image Builder (DIB) and DIB is a modular tool so you can build your own custom workflows. I would expect that DIB has very good RH distro support by default but I have not used RHCOS. Maybe @dtantsur  or @elfosardo is more up to date .

**Comment by agewagra** [comment]
Yes, thank you. Added Simple-init element to get a dhcp less ramdisk with centOS as base image. But even then on the baremetal server, the networkData is not being applied. Am I missing something?

**Comment by Rozzii** [comment]
> Yes, thank you. Added Simple-init element to get a dhcp less ramdisk with centOS as base image. But even then on the baremetal server, the networkData is not being applied.

I have some simple init fixes that I have not pushed upstream because of change in priorities, you could try those: https://github.com/Nordix/glean/commits/parsing_error/ basically glean is the tool that is installed by the simple init element and this takes care of initial network configuration you could install the simple-init and you could export the environment variables listed here https://docs.openstack.org/diskimage-builder/latest/elements/simple-init/README.html to point to my branch and see if it helps.

**Comment by dtantsur** [comment]
/triage accepted
/kind bug
/help

We need to try bringing the fixes to Glean upstream, check with the latest status of the feature in the Ironic project and update the documentation (https://book.metal3.io/bmo/advanced_instance_customization exists but does not cover building the IPA image).
